java/src/main/native/src/RmmJni.cpp

```diff
@@ -43,6 +46,17 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Rmm_shutdown(JNIEnv *env, jclass claz
 
 JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_Rmm_alloc(JNIEnv *env, jclass clazz, jlong size,
                                                       jlong stream) {
+  if (!rmmIsInitialized(nullptr)) {
+    // RMM used to allow allocations without explicit initialization.
```

Comment:
nit: Technically there is a race condition here where multiple alloc calls can happen in parallel.  Right now it is safe because calling rmmInitialize is multiple times does not fail. But we probably want to document the race here.  It would also be good to document in Rmm.java what are the semantic guarantees that it supports.

Reply:
I added synchronization in the allocation codepath when RMM is not initialized so only one thread performs the initialization.  I also updated the javadoc for RMM init and alloc.